File: app/src/main/java/com/example/biteswipe/model/Group.kt
package com.example.biteswipe.model

import java.io.Serializable

/**
 * Data class representing a group in the BiteSwipe app.
 */
data class Group(
    val id: String,
    val name: String,
    val code: String, // Unique code for joining the group
    val creatorId: String, // ID of the user who created the group
    val members: List<String>, // List of user IDs in the group
    val restaurantPreferences: Map<String, List<String>> = emptyMap(), // User ID -> List of liked restaurant IDs
    val matchedRestaurantId: String? = null, // ID of the matched restaurant, if any
    val createdAt: Long, // Timestamp when the group was created
    val expiresAt: Long? = null // Optional timestamp when the group expires
) : Serializable {
    
    /**
     * Checks if all members have submitted their preferences.
     */
    fun allMembersVoted(): Boolean {
        return members.all { memberId -> restaurantPreferences.containsKey(memberId) }
    }
    
    /**
     * Finds the restaurant that has the most likes among group members.
     * Returns the restaurant ID or null if no match is found.
     */
    fun findBestMatch(): String? {
        if (!allMembersVoted()) return null
        
        // Count likes for each restaurant
        val restaurantLikes = mutableMapOf<String, Int>()
        
        restaurantPreferences.values.forEach { likedRestaurants ->
            likedRestaurants.forEach { restaurantId ->
                restaurantLikes[restaurantId] = (restaurantLikes[restaurantId] ?: 0) + 1
            }
        }
        
        // Find the restaurant with the most likes
        return restaurantLikes.entries
            .filter { (_, likes) -> likes > 0 }
            .maxByOrNull { (_, likes) -> likes }
            ?.key
    }
}

File: backend/src/models/Group.ts
import mongoose, { Document, Schema } from 'mongoose';
import crypto from 'crypto';

export interface IGroup extends Document {
  name: string;
  code: string;
  creator: mongoose.Types.ObjectId;
  members: mongoose.Types.ObjectId[];
  restaurantPreferences: Map<string, string[]>; // userId -> [restaurantId, restaurantId, ...]
  matchedRestaurantId?: mongoose.Types.ObjectId;
  createdAt: Date;
  updatedAt: Date;
  expiresAt?: Date;
  allMembersVoted(): boolean;
  findBestMatch(): mongoose.Types.ObjectId | null;
}

const GroupSchema: Schema = new Schema(
  {
    name: {
      type: String,
      required: [true, 'Group name is required'],
      trim: true
    },
    code: {
      type: String,
      required: true,
      unique: true,
      default: () => crypto.randomBytes(3).toString('hex').toUpperCase()
    },
    creator: {
      type: Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    members: [{
      type: Schema.Types.ObjectId,
      ref: 'User'
    }],
    restaurantPreferences: {
      type: Map,
      of: [String]
    },
    matchedRestaurantId: {
      type: Schema.Types.ObjectId,
      ref: 'Restaurant',
      default: null
    },
    expiresAt: {
      type: Date,
      default: () => new Date(Date.now() + 24 * 60 * 60 * 1000) // 24 hours from now
    }
  },
  {
    timestamps: true
  }
);

// Method to check if all members have voted
GroupSchema.methods.allMembersVoted = function(): boolean {
  const memberIds = this.members.map((id: mongoose.Types.ObjectId) => id.toString());
  const votedMemberIds = Array.from(this.restaurantPreferences.keys());
  
  return memberIds.every(memberId => votedMemberIds.includes(memberId));
};

// Method to find the best match
GroupSchema.methods.findBestMatch = function(): mongoose.Types.ObjectId | null {
  if (!this.allMembersVoted()) return null;
  
  // Count likes for each restaurant
  const restaurantLikes = new Map<string, number>();
  
  for (const [_, likedRestaurants] of this.restaurantPreferences.entries()) {
    for (const restaurantId of likedRestaurants) {
      const currentLikes = restaurantLikes.get(restaurantId) || 0;
      restaurantLikes.set(restaurantId, currentLikes + 1);
    }
  }
  
  // Find the restaurant with the most likes
  let bestMatchId: string | null = null;
  let maxLikes = 0;
  
  for (const [restaurantId, likes] of restaurantLikes.entries()) {
    if (likes > maxLikes) {
      maxLikes = likes;
      bestMatchId = restaurantId;
    }
  }
  
  return bestMatchId ? new mongoose.Types.ObjectId(bestMatchId) : null;
};

export default mongoose.model<IGroup>('Group', GroupSchema);